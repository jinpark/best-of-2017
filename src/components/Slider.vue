<template>
  <div class="all">
    <div class="initial">
      <Intro></Intro>
    </div>
    <modal name="youtube" @before-open="beforeOpen" :adaptive=false width="80%" height="80%">
      <YTEmbed :youtubeId="youtubeId"></YTEmbed>
    </modal>
    <swiper :options="swiperOption" ref="mySwiper">
      <!-- slides -->
      <Page v-for="anime in animes" :key="anime.title" v-bind:anime="anime"></Page>
      <End></End>
      <About></About>
      <!-- Optional controls -->
      <div class="swiper-pagination"  slot="pagination"></div>
    </swiper>
  </div>
</template>

<script>
import Intro from './Intro.vue'
import Page from './Page.vue'
import YTEmbed from './YTEmbed.vue'
import End from './End.vue'
import About from './About.vue'
import animes from '../assets/animes.json'

export default {
  name: 'Slider',
  components: {
    Page,
    YTEmbed,
    Intro,
    End,
    About
  },
  methods: {
    beforeOpen (e) {
      this.youtubeId = e.params.youtubeId
    },
    enableScrollEvent (entries, observer) {
      entries.forEach(entry => {
        if (entry.isIntersecting && entry.intersectionRatio > 0.99) {
          entry.target.classList.add('pointer')
        }
      })
    },
    setVideoController () {
      this.swiper.on('slideChange', () => {
        this.$ga.event('slide', 'change', 'normal', this.swiper.activeIndex)
        // pause and play video on slide change
        var activeSlide = this.swiper.slides[this.swiper.activeIndex]
        var previousSlide = this.swiper.slides[this.swiper.previousIndex]
        if (activeSlide.getElementsByTagName('video').length > 0) {
          activeSlide.getElementsByTagName('video')[0].play()
        }
        if (previousSlide.getElementsByTagName('video').length > 0) {
          previousSlide.getElementsByTagName('video')[0].pause()
          previousSlide.getElementsByTagName('video')[0].muted = true
        }
      })
    },
    setObserver () {
      const observer = new window.IntersectionObserver(this.enableScrollEvent, {
        root: null,
        threshold: [0.99]
      })
      observer.observe(document.getElementsByClassName('swiper-container')[0])
    }
  },
  computed: {
    swiper () {
      return this.$refs.mySwiper.swiper
    }
  },
  mounted () {
    this.setVideoController()
    // this.setObserver()
  },
  data () {
    return {
      youtubeId: '',
      swiperOption: {
        direction: 'vertical',
        pagination: {
          el: '.swiper-pagination',
          type: 'bullets'
        },
        mousewheel: true,
        keyboard: true,
        simulateTouch: false
      },
      animes: animes
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.swiper-container {
  width: 100vw;
  height: 100vh;
  pointer-events: none;
}
.initial {
  width: 100vw;
  height: 100vh;
}
.pointer {
  pointer-events: auto;
}

</style>
