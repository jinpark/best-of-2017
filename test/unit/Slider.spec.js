import Slider from '@/components/Slider'
import { shallow } from 'vue-test-utils'
import { createRenderer } from 'vue-server-renderer'

describe('Slider.vue', () => {
  let wrapper
  const animes = [
    {
        "title": "Kizumonogatari (Film Trilogy)",
        "japaneseTitle": "傷物語",
        "filename": "kizumonogatari",
        "summary": "Follows Araragi's encounter with the vampire Kiss-shot Acerola-orion Heart-under-blade and his journey back into humanity from being a vampire.",
        "review": "The Monogatari series has one of the strongest visual style out of all modern anime and with a movie budget, they went all in.",
        "youtubeId": "LYPJoA9udJo"
    },
    {
        "title": "Shouwa Genroku Rakugo Shinjuu",
        "japaneseTitle": "昭和元禄落語心中",
        "filename": "rakugo",
        "summary": "Yotarou is a former yakuza member fresh out of prison and fixated on just one thing: rather than return to a life of crime, the young man aspires to take to the stage of Rakugo, a traditional Japanese form of comedic storytelling.",
        "review": "The first season came out in 2016 but this is one of my favorite shows of all time. This show leans heavily on death and loss.",
        "youtubeId": "rMWOuwFpF-g"
    }
  ]
  Object.defineProperty(window, 'IntersectionObserver', {
    enumerable: false,
    configurable: false,
    writable: true,
    value: jest.fn().mockImplementation(() => {observe: jest.fn()})
  })
  // window.IntersectionObserver.observe = jest.fn()
  var a = new window.IntersectionObserver()
  console.log(a)

  beforeEach(() => {
    wrapper = shallow(Slider, {
      propsData: {animes},
      methods: {
        setVideoController: jest.fn()
      }
    })
  })

  // it('is a swiper-slide', () => {
  //   expect(wrapper.is('swiper-slide')).toBe(true)
  // })

  // it('has correct background video data', () => {
  //   const video = wrapper.find('video')
  //   expect(video.hasAttribute('src', '/static/asilentvoice.mp4')).toBe(true)
  //   expect(video.hasAttribute('poster', '/static/asilentvoice.jpg')).toBe(true)
  // })

  // it('hides mute button by default', () => {
  //   const muteButton = wrapper.find('button.unmute')
  //   const unmuteButton = wrapper.find('button.mute')
  //   expect(muteButton.exists()).toBe(true)
  //   expect(unmuteButton.exists()).toBe(false)
  // })

  // it('toggles mute button when clicked', () => {
  //   expect(wrapper.find('button.unmute').exists()).toBe(true)
  //   expect(wrapper.find('button.mute').exists()).toBe(false)
  //   wrapper.find('button.unmute').trigger('click')
  //   expect(wrapper.find('button.unmute').exists()).toBe(false)
  //   expect(wrapper.find('button.mute').exists()).toBe(true)
  // })

  // it('starts modal when youtube button clicked and mutes all the things', () => {
  //   const youtubeButton = wrapper.find('button.youtube')
  //   youtubeButton.trigger('click')
  //   expect(wrapper.vm.isMuted).toBe(true)
  //   expect(wrapper.vm.video.muted).toBe(true)
  //   expect(wrapper.vm.$modal.show).toHaveBeenCalledWith('youtube', {youtubeId: anime.youtubeId})
  // })

  it('matches snapshot', () => {
    const renderer = createRenderer()
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })
})
