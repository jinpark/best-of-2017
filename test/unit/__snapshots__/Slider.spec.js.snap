// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider.vue matches snapshot 1`] = `
<div class="all">
    <div class="initial">
        <!---->
    </div>
    <modal name="youtube" width="80%" height="80%">
        <!---->
    </modal>
    <swiper options="[object Object]">
        <!---->
        <!---->
        <!---->
        <!---->
        <!---->
        <!---->
        <!---->
        <!---->
        <!---->
        <!---->
        <div slot="pagination" class="swiper-pagination"></div>
    </swiper>
</div>
`;
